xapian-core/api/omenquire.cc

```diff
@@ -467,13 +467,18 @@ MSet::Internal::read_docs() const
 void
 MSet::Internal::set_item_weight(Xapian::doccount i, double wt_)
 {
-    // max_attained is updated as the max of new weights, max_attained of previous weights is discarded.
+    /* This function is called indirectly by MSet::replace_weights through MSet::set_item_weight to update MSet.
+     * This function is indirectly used to assign new weights to the existing documents by MSet::replace_weights.
+     * While assigning new weights max_attained is updated to reflect the max_attained in the new weights.
+     * While assigning new weights max_possible is updated to reflect the max of max_attained and max_possible of previous weights.
+     * Ideally the max_possible should be the maximum possible weight that can be assigned by the weight assigning algorithm
+     * but since it is not practical to calculate max possible weight for different algorithms we use this approach.
```

Comment:
This is fairly verbose (for instance the first two lines say basically the same thing), and still misses the key point: this method _must_ be called in turn on every item in the `MSet` to avoid miscalculation of `max_attained`.

It might be clearer if you put the relevant commentary by the code it refers to. Talk about why the `max_attained` updating works (providing this is called appropriately) next to that code. Talk about why we do `max_possible` like this by the code that implements it. The note about correct calling should certainly be in the docstring for this (internal) method, as well as here, since that's where people are more likely to expect that kind of caveat.

Finally, it is _sometimes_ practical to calculate the `max_possible` weight for different re-sorting algorithms, so you should say something like "but since it may not be possible to calculate…".